File: java/src/main/java/compiler/parser/Parser.java
package compiler.parser;

import compiler.lexer.Scanner;
import compiler.lexer.Token;
import compiler.lexer.TokenType;

public class Parser {
    private Scanner scan;
    private Token currentToken;

    public Parser(byte[] input) {
        this.scan = new Scanner(input);
        this.currentToken = scan.nextToken();
    }

    private void nextToken() {
        this.currentToken = scan.nextToken();
    }

    public void parse() {
        expr();
    }

    // Agora o match verifica o tipo de token, não o lexema
    private void match(TokenType t) {
        if (currentToken.type == t) {
            nextToken();
        } else {
            throw new Error("syntax error");
        }
    }

    // Regra: expr -> number oper
    private void expr() {
        number();
        oper();
    }

    // Método que substitui o antigo digit()
    private void number() {
        System.out.println("push " + currentToken.lexeme);
        match(TokenType.NUMBER);
    }

    // Regra oper validando pelos tipos de token
    private void oper() {
        if (currentToken.type == TokenType.PLUS) {
            match(TokenType.PLUS);
            number();
            System.out.println("add");
            oper();
        } else if (currentToken.type == TokenType.MINUS) {
            match(TokenType.MINUS);
            number();
            System.out.println("sub");
            oper();
        } else if (currentToken.type == TokenType.MULT) {
            match(TokenType.MULT);
            number();
            System.out.println("mult");
            oper();
        } else if (currentToken.type == TokenType.DIV) {
            match(TokenType.DIV);
            number();
            System.out.println("div");
            oper();
        }
    }
}

File: golang/parser/parser.go
package parser

import (
	"fmt"
	"unicode"
)

type Parser struct {
	input   []byte
	current int
}

// Semelhante a um construtor no Java
func NewParser(input []byte) *Parser {
	return &Parser{
		input:   input,
		current: 0,
	}
}

func (p *Parser) Parse() {
	p.expr()
}

func (p *Parser) peek() rune {
	if p.current < len(p.input) {
		return rune(p.input[p.current])
	}
	return '\x00'
}

func (p *Parser) match(c rune) {
	if c == p.peek() {
		p.current++
	} else {
		panic(fmt.Sprintf("syntax error: esperava %c", c))
	}
}

func (p *Parser) expr() {
	p.digit()
	p.oper()
}

func (p *Parser) digit() {
	if unicode.IsDigit(p.peek()) {
		fmt.Printf("push %c\n", p.peek())
		p.match(p.peek())
	} else {
		panic("syntax error: esperava um dígito")
	}
}

func (p *Parser) oper() {
	switch p.peek() {
	case '+':
		p.match('+')
		p.digit()
		fmt.Println("add")
		p.oper()
	case '-':
		p.match('-')
		p.digit()
		fmt.Println("sub")
		p.oper()
	case '*':
		p.match('*')
		p.digit()
		fmt.Println("mult")
		p.oper()
	case '/':
		p.match('/')
		p.digit()
		fmt.Println("div")
		p.oper()
	}
}
